### DISCUSSION TITLE
Allow "pnpm add" command to generate "workspace:*" when adding a workspace dependency

### DISCUSSION THREAD

**Discussion by Mufasa** [pnpm/pnpm#6713] [Ideas]
I have been using npm for many years and have now started to move over to use pnpm. My initial feedback on pnpm is that it is indeed far superior to npm and I absolutely love it. However, I have spotted a strange issue which I have described below.

When I add a workspace dependency using `pnpm add myWorkspaceLibA --workspace --filter myWorkspaceLibB`, pnpm generates this in `myWorkspaceLibB`'s `package.json` file:
```
"dependencies": {
    ...
    "myWorkspaceLibA": "workspace:^",
    ...
  },
```
It also generates this in the root `pnpm-lock.yaml` file:
```
  packages/myWorkspaceLibB:
    dependencies:
      'myWorkspaceLibA':
        specifier: workspace:^
        version: link:../myWorkspaceLibA
```
From the various articles and videos that I have seen on this, I understand that after running the `pnpm add` command I should change the `package.json` entry in `myWorkspaceLibB` to:
```
"dependencies": {
    ...
    "myWorkspaceLibA": "workspace:*",
    ...
  },
```
in order to ensure that it always picks up the latest version of `myWorkspaceLibA` in my monorepo.

However, this leaves the `pnpm-lock.yaml` file inconsistent with the `package.json` file.

I have the following questions on this:

1. Does this matter (i.e. will this cause some unexpected behaviour in some future commands that I may run)?
2. Should I also update the `pnpm-lock.yaml` file to make it consistent with the `package.json` file?

My gut feeling tells me that it is not a good idea for me to be manually changing the `pnpm-lock.yaml` file. I would therefore like to suggest that the `pnpm add` command be enhanced in some way so that it automagically sets `workspace:*` in both the `package.json` and `pnpm-lock.yaml` files.

Is this a reasonable request or have I misunderstood something here?

**Comment by eq-saini-g** [comment]
@Mufasa 
Have you tried using the `--filter` option?

eg: if you have 2 packages in your workspace named `@my-monorepo/pkg-1` and `@my-monorepo/pkg-2`. Then you can do
```bash
pnpm --filter @my-monorepo/pkg-1 add @my-monorepo/pkg-2
```
to add `pkg-2` as a dependency in `pkg-1`

**Reply by Mufasa** [reply]
I am indeed using the `--filter` option as I indicated in `pnpm add myWorkspaceLibA --workspace --filter myWorkspaceLibB`.

The problem is that it adds the dependancy as `"workspace:^"` instead of `"workspace:*"`

**Reply by eq-saini-g** [reply]
Oops!, Silly me. Didn't notice that one earlier.
I'm not aware of the difference between `workspace:^` and `workspace:*`. Can you please clarify that for me?

**Reply by Mufasa** [reply]
`workspace:~` will pickup changes to the patch version but will not pickup any changes to the major or minor versions.

`workspace:^` will pickup changes to the minor or patch versions but will not pickup a change to the major version.

`workspace:*` will pickup changes to the major, minor, or patch versions.

**Reply by eq-saini-g** [reply]
Maybe using the `rolling` option could be of any help - https://pnpm.io/npmrc#save-workspace-protocol

I haven't tried this for my self, so can't say for sure. Please let me know if this works out for you.

**Reply by Mufasa** [reply]
It does not quite meet my needs - I have converted this discussion into an issue where you can find a bit more detail on this here - https://github.com/pnpm/pnpm/issues/6770

**Comment by lightDer** [comment]
I encountered the same problem, and discovered that it can be solved by simply adding double quote to the package name.
e.g `pnpm add "myWorkspaceLibA@workspace:*"`
